src/creature.cpp

```diff
@@ -1042,31 +1039,29 @@ void Creature::onGainExperience(uint64_t gainExp, Creature* target)
 	message.primary.color = TEXTCOLOR_WHITE_EXP;
 	message.primary.value = gainExp;
 
-	for (Creature* spectator : spectators) {
-		assert(dynamic_cast<Player*>(spectator) != nullptr);
-		static_cast<Player*>(spectator)->sendTextMessage(message);
+	for (const auto& spectator : spectators) {
+		assert(spectator->getPlayer() != nullptr);
+		std::static_pointer_cast<Player>(spectator)->sendTextMessage(message);
 	}
 }
 
-bool Creature::setMaster(Creature* newMaster)
+bool Creature::setMaster(std::shared_ptr<Creature> newMaster)
 {
 	if (!newMaster && !master) {
 		return false;
 	}
 
 	if (newMaster) {
-		incrementReferenceCounter();
-		newMaster->summons.push_back(this);
+		// store a shared_ptr reference in the master's summons list to keep the summon alive
+		newMaster->summons.push_back(getCreature());
 	}
 
-	Creature* oldMaster = master;
-	master = newMaster;
+	std::swap(master, newMaster);
 
-	if (oldMaster) {
-		auto summon = std::find(oldMaster->summons.begin(), oldMaster->summons.end(), this);
-		if (summon != oldMaster->summons.end()) {
-			oldMaster->summons.erase(summon);
-			decrementReferenceCounter();
+	if (newMaster) {
+		auto summon = std::find(newMaster->summons.begin(), newMaster->summons.end(), getCreature());
```

Comment:
Could this be `const auto`
```suggestion
		const auto summon = std::find(newMaster->summons.begin(), newMaster->summons.end(), getCreature());
```